src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs

```diff
@@ -90,15 +90,37 @@ public override bool IsInvalid
 
         // SafeHandle abstract member
         protected override bool ReleaseHandle()
+        {
+            Dispose(disposing: false);
+            return true;
+        }
+
+        protected override void Dispose(bool disposing)
         {
             if (!Disposed)
             {
+                base.Dispose(disposing);
+                if (!disposing)
+                {
+                    Disposed = true;
+                    VirtualMemoryManager.VirtualFreeEx(ProcessHandle,
+                        AllocatedPointer, MemorySize);
+                }
                 Disposed = true;
-                GC.SuppressFinalize(this);
-                VirtualMemoryManager.VirtualFreeEx(ProcessHandle,
-                    AllocatedPointer, MemorySize);
             }
-            return true;
+        }
```

Comment:
This class is deriving from SafeHandle, so it is a bit of a special case in terms of memory clean up. Because the base class implements the IDisposable pattern, it exposes the regular Dispose(bool) method that you can override to clean up. In addition, the abstract base class exposes the ReleaseHandle method (which is also designed for cleanup). Making things just a little bit harder, if you dig into the base dispose method you will find that it ends in some extern calls that get implemented by the runtime (making it tricky to tell what is actually happening). Lot of details in these docs here: SafeHandle Class (System.Runtime.InteropServices) | Microsoft Docs. Take note of the section at the very bottom titled:  Notes to Implementers 
What is convenient here is we can assume that MS has implemented the dispose correctly, so the amount of code we need to write is significantly reduced. Let’s start by looking at the regular Dispose(bool) (I am just using the default VS generates). This will be functionally similar to what SafeHandle will do:
```C#
protected virtual void Dispose(bool disposing)
{
    if (!_disposedValue)
    {
        if (disposing)
        {
            // TODO: dispose managed state (managed objects)
        }

        // TODO: free unmanaged resources (unmanaged objects) and override finalizer
        // TODO: set large fields to null
        _disposedValue = true;
    }
}
```
The Dispose implementation inside of SafeHandle will ensure that Dispose gets appropriately called (passing the right disposing value). In the TODO section for freeing unmanaged resources is where it will call ReleaseHandle (expecting that it will do the needful to cleanup unmanaged handles). You will also notice that it does this regardless of whether the disposing flag. So for our code we only need to implement ReleaseHandle, and don’t need to do any of the regular disposable members as those are already handled for us.
So for our code here the implementation we want should look like this:
```C#
protected override bool ReleaseHandle()
{
    Disposed = true;
    VirtualMemoryManager.VirtualFreeEx(ProcessHandle, AllocatedPointer, MemorySize);
}
```